src/main/java/com/dnd/reetplace/app/service/OAuth2Service.java

```diff
@@ -0,0 +1,121 @@
+import com.dnd.reetplace.app.dto.auth.KakaoProfileResponse;
+import com.dnd.reetplace.app.dto.auth.LoginResponse;
+import com.dnd.reetplace.app.dto.auth.RefreshTokenDto;
+import com.dnd.reetplace.app.dto.auth.TokenResponse;
+import com.dnd.reetplace.app.repository.MemberRepository;
+import com.dnd.reetplace.app.type.LoginType;
+import com.dnd.reetplace.global.exception.auth.KakaoUnauthorizedException;
+import com.dnd.reetplace.global.security.TokenProvider;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestTemplate;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.Map;
+
+import static org.springframework.http.HttpHeaders.AUTHORIZATION;
+
+@Slf4j
+@Transactional(readOnly = true)
+@SuppressWarnings("unchecked")
+@RequiredArgsConstructor
+@Service
+public class OAuth2Service {
+
+    public static final String GET_PROFILE_URL = "https://kapi.kakao.com/v2/user/me";
+    private final MemberRepository memberRepository;
+    private final RefreshTokenRedisService refreshTokenRedisService;
+    private final TokenProvider tokenProvider;
+
+    /**
+     * 카카오 서버에서 받은 access token을 기반으로 사용자 프로필을 받아온 후, 로그인 (또는 회원가입)을 진행한다.
+     *
+     * @param request Authorization Header(kakao access token) 가 포함된 HttpServletRequest 객체
+     * @return 로그인 (또는 회원가입) 완료 후 사용자 정보 (memberId, uid, accessToken, refreshToken 등)
+     */
+    @Transactional
+    public LoginResponse kakaoLogin(HttpServletRequest request) throws JsonProcessingException {
+
+        String token = tokenProvider.getToken(request);
+
+        // Header 추가
+        HttpHeaders header = new HttpHeaders();
+        header.add(AUTHORIZATION, "Bearer " + token);
+        header.add("Content-type", "application/x-www-form-urlencoded;charset=utf-8");
+
+        // request 구성
+        RestTemplate rt = new RestTemplate();
+        rt.setRequestFactory(new HttpComponentsClientHttpRequestFactory());
+        HttpEntity<MultiValueMap<String, String>> httpRequest = new HttpEntity<>(header);
+
+        // API 요청
+        ResponseEntity<String> response;
+        try {
+            response = rt.exchange(
+                    GET_PROFILE_URL,
+                    HttpMethod.POST,
+                    httpRequest,
+                    String.class
+            );
+        } catch (Exception e) {
+            log.error("OAuth2Service.kakaoLogin() ex={}", String.valueOf(e));
```

Comment:
로그를 출력할 때 어떤 요청에 대한 로그인지를 명시하는게 좋을 것 같습니다.
다음과 같이 한다면 각 요청에 대해 고유하게 생성된 log trace id를 함께 출력할 수 있습니다.
```suggestion
            log.error("[{}] OAuth2Service.kakaoLogin() ex={}", LogUtils.getLogTraceId(), String.valueOf(e));
```

사실 가능하다면 모든 로그 출력에 해당 prefix를 자동으로 넣고 싶은데... 시간이 없어서(귀찮아서..😅) 아직 가능한 방법인지 확인을 못해봤네요.